### PR TITLE
Feature/browser redirect

### PR DESCRIPTION
**Pull request by gargravarr**
- Finished separating functionality of zign and ztoken
- Still no Unit Tests for ztoken, will write them on a next PR

### CODE REVIEW

File: zign/api.py [hunk 1]
```diff
@@ -6,12 +6,87 @@
...
+                window.location.href = "http://localhost:{port}/?" + fragment
+            }} else {{
+                document.getElementById("error").style = "display: block;"
+            }}
+        }})();
+    </script>
+  </head>
+  <body>
+    <p style="display:none;" id="error">
+        Please put a nice message here.
```
Comment (hjacobs) [inline zign/api.py]: so we should put a nice message there I guess :smirk:

Comment (hjacobs) [reply zign/api.py]: We should also have a `<noscript>` section for people having JavaScript disabled!

File: zign/api.py [hunk 2]
```diff
@@ -34,8 +109,69 @@ def __str__(self):
...
+        else:
+            self.server.query_params = parse_qs(query_string)
+            if 'access_token' in self.server.query_params:
+                page = SUCCESS_PAGE
+            else:
+                page = ERROR_PAGE
+            self.wfile.write(page.encode('utf-8'))
+
+
+def get_config(config_module=None, override=dict()):
```
Comment (hjacobs) [inline zign/api.py]: never use `dict()` as default parameter in Python --- it's only initialized once, so you are operating on a singleton!

Better: 

```
def foo(a=None):
  a = a or {} # use empty dict if a is None or empty

File: zign/api.py [hunk 3]
```diff
@@ -93,21 +229,91 @@ def store_token(name: str, result: dict):
...
+                break
+            port_number += 1
+        else:
+            success = True
+            break
+
+    if success:
+        params = {'response_type':          'token',
+                  'scope':                  config['scope'],
+                  'business_partner_id':    config['partner_id'],
```
Comment (hjacobs) [inline zign/api.py]: shouldn't we use `business_partner_id` in config too (for consistency)?

File: zign/cli.py [hunk 4]
```diff
@@ -40,15 +42,17 @@ def format_expires(token: dict):
...
-@click.pass_obj
-def list_tokens(obj, output):
+def list_tokens(output):
     '''List tokens'''
     data = get_tokens()
 
     rows = []
     for key, val in sorted(data.items()):
+        access_token = val.get('access_token')
+        if output == 'text' and len(access_token) > 36:
```
Comment (hjacobs) [inline zign/cli.py]: better use the column width feature of `print_table` instead of truncating manually here

File: zign/cli.py [hunk 5]
```diff
@@ -75,27 +79,21 @@ def delete_token(obj, name):
...
 @cli.command()
-@click.argument('scope', nargs=-1)
-@click.option('--url', help='URL to generate access token', metavar='URI')
-@click.option('--realm', help='Use custom OAuth2 realm', metavar='NAME')
 @click.option('-n', '--name', help='Custom token name (will be stored)', metavar='TOKEN_NAME')
-@click.option('-U', '--user', help='Username to use for authentication', envvar='ZIGN_USER', metavar='NAME')
-@click.option('-p', '--password', help='Password to use for authentication', envvar='ZIGN_PASSWORD', metavar='PWD')
-@click.option('--insecure', help='Do not verify SSL certificate', is_flag=True, default=False)
+@click.option('-s', '--scope', help='Scope to request access to', metavar='SCOPE')
+@click.option('-a', '--auth_url', help='Auth URL to generate access token', metavar='AUTH_URL')
```
Comment (hjacobs) [inline zign/cli.py]: please use `--auth-url` (dashes in options will automatically be converted to underscore for Python params)

File: zign/cli.py [hunk 6]
```diff
@@ -75,27 +79,21 @@ def delete_token(obj, name):
...
+@click.option('-c', '--client_id', help='Client ID to use', metavar='CLIENT_ID')
```
Comment (hjacobs) [inline zign/cli.py]: same here: `--client-id`

File: zign/cli.py [hunk 7]
```diff
@@ -75,27 +79,21 @@ def delete_token(obj, name):
...
+@click.option('-p', '--partner_id', help='Business Partner ID to use', metavar='PARTNER_ID')
```
Comment (hjacobs) [inline zign/cli.py]: `--business-partner-id` for consitency? (short form `-p` can stay)

**Comment by hjacobs** [comment]
@gargravarr we should not have a deprecation message for Zign yet --- we want to roll it out in "shadow mode" without disturbing the users. We can add a deprecation message in a later release.

File: zign/cli.py [hunk 8]
```diff
@@ -40,22 +42,22 @@ def format_expires(token: dict):
...
-                     'access_token': val.get('access_token'),
+                     'access_token': access_token,
                      'scope': val.get('scope'),
                      'creation_time': val.get('creation_time'),
                      'expires_in': format_expires(val)})
 
     with OutputFormat(output):
         print_table('name access_token scope creation_time expires_in'.split(), rows,
-                    titles={'creation_time': 'Created'})
+                    titles={'creation_time': 'Created'}, max_column_widths=36)
```
Comment (hjacobs) [inline zign/cli.py]: AFAIK max_column_widths needs to be a map col name -> col width

**Comment by gargravarr** [comment]
:+1:

**Comment by gargravarr** [comment]
@prayerslayer can you take a look at [95b0ad](https://github.com/zalando-stups/zign/pull/17/commits/95b0daddbfe64f9c8f3c524334a4b166fe4363fb)?

File: zign/api.py [hunk 9]
```diff
@@ -66,6 +66,11 @@
             }}
         }})();
     </script>
+    <noscript>
```
Comment (prayerslayer) [inline zign/api.py]: I think it's a good idea to do this, however:

* Multiple nodes with same ID (`error`) in this document => undefined behavior. Don't do it.
* `noscript` in `head` is not allowed to contain flow content such as `<p>`, put it in `<body>`.
* What's the point of `display:none`? `noscript` content will only be inserted when scripting is disabled.

Suggestion:

~~~ html
<body>
  <noscript>
    <p>Your browser does not support Javascript! Please enable it or switch to a Javascript enabled browser.</p>
  </noscript>
  <!-- actual content -->
</body>
~~~

**Comment by hjacobs** [comment]
We should change the config to clearly indicate what `auth_url` is, so I propose renaming it to `authorize_url` as it's the URL for the "Authorization Endpoint". It's also called like this in Flask OAuth and oauth2client itself: https://github.com/zalando-incubator/oauth2-proxy/blob/master/oauth2_proxy/app.py#L60

**Comment by hjacobs** [comment]
I would also propose changing the config dir to "zalando-token-cli" as I would use that as package and repo name in the future ("ztoken" being the short form as console script)

**Comment by hjacobs** [comment]
:+1:

File: zign/api.py [hunk 10]
```diff
@@ -93,21 +231,90 @@ def store_token(name: str, result: dict):
...
+        parsed_authorize_url = urlparse(config['authorize_url'])
+        browser_url = urlunsplit((parsed_authorize_url.scheme, parsed_authorize_url.netloc, parsed_authorize_url.path,
+                                  param_string, ''))
+
+        webbrowser.open(browser_url, new=1, autoraise=True)
+        click.echo('Your browser has been opened to visit:\n\n\t{}\n'.format(browser_url))
+    else:
+        raise AuthenticationFailed('Failed to launch local server')
+
+    # Handle first request, which will redirect to Javascript
```
Comment (hjacobs) [inline zign/api.py:284]: this is a bit brittle in my eyes

**Pull request merged**